### PR TITLE
Split ZVertexSoA in a run-time sized PortableMultiCollection

### PR DESCRIPTION
**Pull request by fwyzard**
#### PR description:

Split `ZVertexSoA` in two layouts, and wrap them in a multi-layout collection.
Make the ZVertex SoA collections run-time sized, and replace the `ZVertexHost`/`ZVertexDevice` template classes inheriting from `PortableHostMultiCollection`/`PortableDeviceMultiCollection` with type aliases.

Remove versioning info from class description for dictionaries.
Port comments from the CUDA version.

The maximum number of vertices is now a configuration parameter of the `PixelVertexProducerAlpaka` `EDProducer`, while all device-side checks now make use of the various SoA capacity.
The maximum number of vertices is set to
  - `256` for Run 3
  - `512` for Phase 2
  - `32` for heavy ions
These values are probably conservative, and could be further reduced.
The tests use the legacy hard-coded values, taken from the old Run 3 defaults.
    
The documentation in `DataFormats/VertexSoA/README.md` is updated accordingly.

#### PR validation:

Unit tests and workflow `12834.403` pass.

### CODE REVIEW

**Review by makortel — COMMENTED** [review]
Just to make it clear (or remind), this PR will make all files with the earlier definition of `ZVertexHost` unreadable (which I believe to be fine).

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
@AdrianoDee can you double check this, and eventually implement the necessary configuration changes for the Phase 2 workflows ?

**Comment by fwyzard** [comment]
enable gpu

**Comment by fwyzard** [comment]
please test with #45694

**Comment by fwyzard** [comment]
please test

**Comment by AdrianoDee** [comment]
> can you double check this, and eventually implement the necessary configuration changes for the Phase 2 workflows ?

Sure

File: RecoTracker/PixelVertexFinding/plugins/alpaka/PixelVertexProducerAlpaka.cc [hunk 1]
```diff
@@ -61,6 +64,7 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE {
               conf.getParameter<double>("eps"),
               conf.getParameter<double>("errmax"),
               conf.getParameter<double>("chi2max")),
+        maxVertices_(conf.getParameter<int>("maxVertices")),
```
Comment (AdrianoDee) [inline RecoTracker/PixelVertexFinding/plugins/alpaka/PixelVertexProducerAlpaka.cc:67]: Ii think we can lower the default (Run3) here to something like `256` (if we want to keep powers of 2, `128` could happen for some very high PU event) given we have

image

[in a recent TTbar PU 2024 RelVal](https://tinyurl.com/2ars67bb). Then we could have:

- `512` for Phase2;
- `32` for HIN;

I've put the changes in https://github.com/AdrianoDee/cmssw/commit/a28e5490b62e03ca97a904005ad9277cc332c995.

Comment (AdrianoDee) [reply RecoTracker/PixelVertexFinding/plugins/alpaka/PixelVertexProducerAlpaka.cc:67]: The `1024` was there to be super safe for Phase2. Maybe too much safe.

**Comment by mmusich** [comment]
This needs now a rebase (because https://github.com/cms-sw/cmssw/pull/45842 was merged in the meanwhile)

File: RecoTracker/PixelVertexFinding/plugins/alpaka/PixelVertexProducerAlpaka.cc (hunk 1 above)
Comment (fwyzard) [reply RecoTracker/PixelVertexFinding/plugins/alpaka/PixelVertexProducerAlpaka.cc:67]: It doesn't need to be a power of 2 :-)

**Comment by fwyzard** [comment]
> This needs now a rebase (because #45842 was merged in the meanwhile)

OK, I'll come back to this once we have an IB with it.

**Comment by fwyzard** [comment]
Rebased on top of a recent IB, and cherry-picked @AdrianoDee's update to the collection sizes.

**Comment by fwyzard** [comment]
enable gpu

**Comment by fwyzard** [comment]
test parameters:
  - workflows = 12834.403
  - enable_tests = gpu
  - workflows_gpu = 12834.403

**Comment by fwyzard** [comment]
please test

**Comment by AdrianoDee** [comment]
test parameters:
- enable_test = gpu

**Comment by AdrianoDee** [comment]
please test

Something convoluted happened: ˋ12834.403ˋ is now included in the default gpu tests (and indeed has been run). Then it's "not found" as an extra wf since it would need `relval_opt_gpu = -w upgrade`. 🪄

**Comment by AdrianoDee** [comment]
please abort

**Comment by AdrianoDee** [comment]
please test

**Comment by fwyzard** [comment]
can you ask the bot to test it also on cpu ?

**Comment by AdrianoDee** [comment]
please abort

**Comment by AdrianoDee** [comment]
test parameters:

- workflows = 12834.403
- enable_tests = gpu
- relval_opts = -w upgrade

**Comment by AdrianoDee** [comment]
please abort

**Comment by fwyzard** [comment]
+heterogeneous

**Comment by mmusich** [comment]
+hlt
   * (direct) changes in HLT domain are minimal (`RecoTauTag/HLTProducers/src/L2TauTagNNProducerAlpaka.cc`)

**Comment by fwyzard** [comment]
kind ping to @cms-sw/dqm-l2 🙏🏻

**Comment by antoniovagnerini** [comment]
+1
- spurious differences seen from unstable WF `12434.7`  and `12834.7`   (open issue #39803 )

**Comment by mandrenguyen** [comment]
+1

**Comment by mandrenguyen** [comment]
cross posting the following comment:
https://github.com/cms-sw/cmssw/pull/46686#issuecomment-2473637407

Is there a failure related to this PR?

**Comment by makortel** [comment]
> Is there a failure related to this PR?

Nothing has been reported since the PR was merged, and the IB GPU tests are clean (for the history that is visible). On the other hand, this doesn't prove much because the workflow that failed in https://github.com/cms-sw/cmssw/pull/46686#issuecomment-2473565507, 14949.402, is not tested in IBs.

**Pull request merged**